### DISCUSSION TITLE
java.lang.NoClassDefFoundError: javax/xml/bind/JAXBContext

### DISCUSSION THREAD

**Discussion by imranrajjad** [alapierre/ksef-java-rest-client#110] [Q&A] [answered]
I am trying out the sample code to run Authentication using self signed p12 file, just as demonstrated, however my code is failing and complaining about missing Jaxb context, I have included the correct JAXB files in my pom

```
<dependency>
        <groupId>jakarta.xml.bind</groupId>
        <artifactId>jakarta.xml.bind-api</artifactId>
        <version>4.0.1</version>
    </dependency>
    <dependency>
        <groupId>org.glassfish.jaxb</groupId>
        <artifactId>jaxb-runtime</artifactId>
        <version>4.0.4</version>
    </dependency>
```

If anyone can confirm if this is the correct version

Also this is how I am trying to login

```
 P12Signer signer = new P12Signer(pas, tokenFile);
        KsefDssFacade facade = new KsefDssFacade(signer, sesjaApi);
        return facade.authByDigitalSignature(NIP_FIRMY, IdentifierType.onip);
```

Error static from my unit test

```
D:\einvoice\vs_work\ksef\target\test-classes\ktnassertprod.p1210:39:32.688 [main] DEBUG io.alapierre.ksef.api.dss.facade.KsefDssFacade - challengeResponse = AuthorisationChallengeResponse(timestamp=2023-12-31T07:39:31.723Z, challenge=20231231-CR-D4A3B3F495-9284F01471-98)
10:39:33.970 [main] DEBUG io.alapierre.crypto.dss.signer.AbstractSigner - sprawdzam certyfikat Mon Sep 11 01:47:51 AST 2023 Tue Sep 10 01:47:51 AST 2024
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.018 s <<< FAILURE! - in com.lst.ksef.AppTest
[ERROR] shouldAnswerWithTrue(com.lst.ksef.AppTest)  Time elapsed: 2.968 s  <<< ERROR!
java.lang.NoClassDefFoundError: javax/xml/bind/JAXBContext
```
As for the Token based Authentication, it is quite unclear how to start a session when you only have NIP number as I get a BAD Request ERROR, really struggling with documentation here.

thanks

**Comment by alapierre** [comment]
hallo,

thank you for your message. The problem is in digidoc4j version - I used 5.13.RC1 in last KSeF client. Now there is stable 6.0 and it works correct with JAXB 4.x. In next few days I will release crypto-util project and update dependencies in KSeF Client. 

I do not understand your question about token auth - did you check Ministry of Finance documentation? You can look at samples here: https://github.com/ksef4dev/ksef-java-sample I hope it will helps.

**Reply by imranrajjad** [reply]
@alapierre thanks you your response, Is there a way to validate the Invoivce payload before sending it to REST End? For the purpose of marshalling the XML into its respective Java class and maybe performing any modifications or implementing fallback mechanisms to ensure a Valid XML always goes out? Or do you recommend using the XSD files to generate them separately when using these libraries?

**Reply by alapierre** [reply]
you can validate XML against schema, eg. you can use this util: https://github.com/alapierre/java-commons/tree/master/xml-utils `XMLValidator`

It is important to check invoice duplicates on your side, because MF check it only on the same session, so it is possible to send the same invoice more than once.

**Comment by alapierre** [comment, marked as answer]
New version of KSeF client (`2.0.24`) is ready. I prepared small sample for digital signature login here: https://github.com/ksef4dev/digital-stamp-sample
